fast_llm/data/dataset/gpt/memmap.py

```
@@ -116,13 +160,45 @@ def get(
         sample_spans = None
         if use_loss_masking_spans and self._spans is not None:
             sample_spans = self._spans[idx]
-            # adjust the spans for the offset and length
+
+            # filter spans that are outside the range of the selected tokens in the document
             sample_spans = sample_spans[
                 (sample_spans[:, 0] < offset + len(token_ids)) & (sample_spans[:, 1] >= offset)
             ]
-            sample_spans[:, 0] = np.maximum(sample_spans[:, 0], offset) - offset
+
+            # subtract by offset to normalize span boundaries
+            sample_spans[:, 0] = np.maximum(sample_spans[:, 0], offset) - offset  # offset
             sample_spans[:, 1] = np.minimum(sample_spans[:, 1], offset + len(token_ids) - 1) - offset
-        return GPTSample(token_ids=token_ids, loss_masking_spans=sample_spans)
+
+        chosen_span = None
+        rejected_span = None
+        if use_preference_loss_masking_spans and self._chosen_spans is not None and self._rejected_spans is not None:
```

is there any scenario where `(self._chosen_spans is not None) != (self._rejected_spans is not None)`

---

Also given the intended usage we really want both to exist when `use_preference_loss_masking_spans=True`, otherwise DPO will crash...

---

In terms of expected behavior I think both chosen spans and rejected spans should always be specified.  I've added a few extra checks so that it fails early if either of these are empty when preference spans are enabled.